File: scripts/taxonomy.py
#!/usr/bin/env python3
"""
Taxonomy and Watchlist Configuration for Google Cloud Radar.

Hierarchical 3-level taxonomy:
1. Ecosystems: List of ecosystems with categories ID list
2. Categories: List of categories with services ID list
3. Services: Dictionary of services with canonical names, aliases, and release feeds

Re-/moving service happens by updating the parent category's services list.
Changing category/ecosystem metadata happens within their respective objects.
"""

import json
import re
from pathlib import Path
from typing import TypedDict, Optional


class ServiceMeta(TypedDict, total=False):
    ecosystem: str
    ecosystem_id: str
    category: str
    category_id: str
    name: str
    aliases: list[str]
    release_feed_url: str
    release_feed_urls: list[str]


TAXONOMY_PATH = Path(__file__).resolve().parent.parent / "data" / "taxonomy.json"

with open(TAXONOMY_PATH, "r", encoding="utf-8") as _f:
    _TAXONOMY_DATA = json.load(_f)

_ECOSYSTEMS_RAW: list[dict] = _TAXONOMY_DATA.get("ecosystems", [])
_CATEGORIES_RAW: list[dict] = _TAXONOMY_DATA.get("categories", [])
_SERVICES_RAW: dict[str, dict] = _TAXONOMY_DATA.get("services", {})

ECOSYSTEMS_BY_ID: dict[str, dict] = {e["id"]: e for e in _ECOSYSTEMS_RAW}
CATEGORIES_BY_ID: dict[str, dict] = {c["id"]: c for c in _CATEGORIES_RAW}

# Parent mappings
CATEGORY_TO_ECOSYSTEM: dict[str, dict] = {}
for eco in _ECOSYSTEMS_RAW:
    for cat_id in eco.get("categories", []):
        CATEGORY_TO_ECOSYSTEM[cat_id] = eco

SERVICE_TO_CATEGORY: dict[str, dict] = {}
for cat in _CATEGORIES_RAW:
    for svc_id in cat.get("services", []):
        SERVICE_TO_CATEGORY[svc_id] = cat

SERVICE_TO_ECOSYSTEM: dict[str, dict] = {}
for eco in _ECOSYSTEMS_RAW:
    for svc_id in eco.get("services", []):
        SERVICE_TO_ECOSYSTEM[svc_id] = eco

# Resolved ServiceMeta for all services
WATCHED_SERVICES: dict[str, ServiceMeta] = {}
for svc_id, svc_data in _SERVICES_RAW.items():
    cat = SERVICE_TO_CATEGORY.get(svc_id)
    if cat:
        eco = CATEGORY_TO_ECOSYSTEM.get(cat.get("id"), {})
        category_name = cat.get("name", "")
        category_id = cat.get("id", "")
    else:
        eco = SERVICE_TO_ECOSYSTEM.get(svc_id, {})
        category_name = ""
        category_id = ""

    meta: ServiceMeta = {
        "name": svc_data.get("name", svc_id),
        "ecosystem": eco.get("name", "More"),
        "ecosystem_id": eco.get("id", "more"),
        "aliases": svc_data.get("aliases", []),
        "release_feed_url": svc_data.get("release_feed_url"),
        "release_feed_urls": svc_data.get("release_feed_urls", []),
    }
    if category_name:
        meta["category"] = category_name
        meta["category_id"] = category_id

    WATCHED_SERVICES[svc_id] = meta

# Backward compatibility lists
ECOSYSTEMS: list[str] = [e["name"] for e in _ECOSYSTEMS_RAW]
CATEGORIES: list[str] = [c["name"] for c in _CATEGORIES_RAW]


def get_ecosystems() -> list[dict]:
    """Returns the list of all ecosystem definitions with categories ID list."""
    return _ECOSYSTEMS_RAW


def get_categories() -> list[dict]:
    """Returns the list of all category definitions with services ID list."""
    return _CATEGORIES_RAW


def get_services() -> dict[str, dict]:
    """Returns the raw service definitions map."""
    return _SERVICES_RAW


def _find_service_meta(service_or_api: str) -> Optional[ServiceMeta]:
    """Finds matching ServiceMeta using exact key, API prefix, aliases, or name."""
    if not service_or_api:
        return None
    lower = service_or_api.lower().strip()
    if lower in WATCHED_SERVICES:
        return WATCHED_SERVICES[lower]

    # Check API prefix (e.g. "aiplatform.v1" -> "aiplatform", "discoveryengine:v1" -> "discoveryengine")
    prefix = lower.split(".")[0].split(":")[0]
    if prefix in WATCHED_SERVICES:
        return WATCHED_SERVICES[prefix]

    clean = re.sub(r"[^a-z0-9]", "", lower)
    if clean in WATCHED_SERVICES:
        return WATCHED_SERVICES[clean]

    clean_prefix = re.sub(r"[^a-z0-9]", "", prefix)
    if clean_prefix in WATCHED_SERVICES:
        return WATCHED_SERVICES[clean_prefix]

    # Exact or alias/name matches
    for key, meta in WATCHED_SERVICES.items():
        name = meta.get("name", "").lower()
        aliases = [a.lower() for a in meta.get("aliases", []) if isinstance(a, str)]
        for candidate in [name] + aliases:
            if not candidate:
                continue
            if candidate == lower:
                return meta
            clean_cand = re.sub(r"[^a-z0-9]", "", candidate)
            if clean and clean == clean_cand:
                return meta

    # Check if a full canonical name or alias is contained in the query string
    for key, meta in WATCHED_SERVICES.items():
        name = meta.get("name", "").lower()
        aliases = [a.lower() for a in meta.get("aliases", []) if isinstance(a, str)]
        for candidate in [name] + aliases:
            if candidate and len(candidate) >= 4 and candidate in lower:
                return meta

    return None


def is_watched_api(api_name: str) -> bool:
    """Returns True if the API is on the watchlist."""
    return _find_service_meta(api_name) is not None


def get_watched_api_names() -> list[str]:
    """Returns a sorted list of all watched API names."""
    return sorted(WATCHED_SERVICES.keys())


def get_ecosystem_for_service(service_or_api: str) -> str:
    """Returns top-level ecosystem name (e.g. AI/ML, Databases and analytics, etc.)."""
    meta = _find_service_meta(service_or_api)
    if meta:
        return meta.get("ecosystem", "More")
    return "More"


def get_category_for_service(service_or_api: str) -> Optional[str]:
    """Classifies a service or API name into a standardized category if one exists."""
    meta = _find_service_meta(service_or_api)
    if meta:
        return meta.get("category")
    return None



def get_release_feed_urls(service_or_api: str) -> list[str]:
    """Returns list of official Google release RSS/Atom feed URLs for a service if configured."""
    meta = _find_service_meta(service_or_api)
    if not meta:
        return []

    urls = []
    if "release_feed_urls" in meta and isinstance(meta["release_feed_urls"], list):
        for u in meta["release_feed_urls"]:
            if u and u not in urls:
                urls.append(u)
    if "release_feed_url" in meta and meta["release_feed_url"]:
        if meta["release_feed_url"] not in urls:
            urls.append(meta["release_feed_url"])
    return urls


def get_release_feed_url(service_or_api: str) -> Optional[str]:
    """Returns the primary official Google release RSS/Atom feed URL for backward compatibility."""
    urls = get_release_feed_urls(service_or_api)
    return urls[0] if urls else None


def get_official_release_feeds() -> dict[str, str]:
    """Returns a mapping of service identifier to official release notes RSS feed URL."""
    feeds = {}
    for key in WATCHED_SERVICES:
        urls = get_release_feed_urls(key)
        if urls:
            feeds[key] = urls[0]
    return feeds

File: scripts/taxonomy.ts
/**
 * Taxonomy and Watchlist Configuration for Google Cloud Radar.
 *
 * Hierarchical 3-level taxonomy:
 * 1. Ecosystems: List of ecosystems with categories ID list
 * 2. Categories: List of categories with services ID list
 * 3. Services: Dictionary of services with canonical names, aliases, and release feeds
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

export interface ServiceMeta {
  name: string;
  ecosystem: string;
  ecosystem_id: string;
  category?: string;
  category_id?: string;
  aliases?: string[];
  release_feed_url?: string;
  release_feed_urls?: string[];
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const TAXONOMY_PATH = path.resolve(__dirname, '../data/taxonomy.json');

const taxonomyRaw = JSON.parse(fs.readFileSync(TAXONOMY_PATH, 'utf-8'));

const _ECOSYSTEMS_RAW: Array<{ id: string; name: string; categories?: string[]; services?: string[] }> = taxonomyRaw.ecosystems || [];
const _CATEGORIES_RAW: Array<{ id: string; name: string; services?: string[] }> = taxonomyRaw.categories || [];
const _SERVICES_RAW: Record<string, any> = taxonomyRaw.services || {};

export const ECOSYSTEMS_BY_ID: Record<string, any> = Object.fromEntries(_ECOSYSTEMS_RAW.map((e) => [e.id, e]));
export const CATEGORIES_BY_ID: Record<string, any> = Object.fromEntries(_CATEGORIES_RAW.map((c) => [c.id, c]));

// Parent mappings
const CATEGORY_TO_ECOSYSTEM: Record<string, any> = {};
for (const eco of _ECOSYSTEMS_RAW) {
  for (const catId of eco.categories || []) {
    CATEGORY_TO_ECOSYSTEM[catId] = eco;
  }
}

const SERVICE_TO_CATEGORY: Record<string, any> = {};
for (const cat of _CATEGORIES_RAW) {
  for (const svcId of cat.services || []) {
    SERVICE_TO_CATEGORY[svcId] = cat;
  }
}

const SERVICE_TO_ECOSYSTEM: Record<string, any> = {};
for (const eco of _ECOSYSTEMS_RAW) {
  for (const svcId of eco.services || []) {
    SERVICE_TO_ECOSYSTEM[svcId] = eco;
  }
}

// Resolved ServiceMeta for all services
export const WATCHED_SERVICES: Record<string, ServiceMeta> = {};
for (const [svcId, svcData] of Object.entries(_SERVICES_RAW)) {
  const cat = SERVICE_TO_CATEGORY[svcId];
  let eco: any = {};
  let categoryName = '';
  let categoryId = '';

  if (cat) {
    eco = CATEGORY_TO_ECOSYSTEM[cat.id] || {};
    categoryName = cat.name || '';
    categoryId = cat.id || '';
  } else {
    eco = SERVICE_TO_ECOSYSTEM[svcId] || {};
  }

  const meta: ServiceMeta = {
    name: svcData.name || svcId,
    ecosystem: eco.name || 'More',
    ecosystem_id: eco.id || 'more',
    aliases: svcData.aliases || [],
    release_feed_url: svcData.release_feed_url,
    release_feed_urls: svcData.release_feed_urls || [],
  };

  if (categoryName) {
    meta.category = categoryName;
    meta.category_id = categoryId;
  }

  WATCHED_SERVICES[svcId] = meta;
}

export const ECOSYSTEMS: string[] = _ECOSYSTEMS_RAW.map((e) => e.name);
export const CATEGORIES: string[] = _CATEGORIES_RAW.map((c) => c.name);

export function getEcosystems(): any[] {
  return _ECOSYSTEMS_RAW;
}

export function getCategories(): any[] {
  return _CATEGORIES_RAW;
}

export function getServices(): Record<string, any> {
  return _SERVICES_RAW;
}

export function findServiceMeta(serviceOrApi: string): ServiceMeta | null {
  if (!serviceOrApi) return null;
  const lower = serviceOrApi.toLowerCase().trim();
  if (WATCHED_SERVICES[lower]) {
    return WATCHED_SERVICES[lower];
  }

  // Check API prefix (e.g. "aiplatform.v1" -> "aiplatform", "discoveryengine:v1" -> "discoveryengine")
  const prefix = lower.split('.')[0].split(':')[0];
  if (WATCHED_SERVICES[prefix]) {
    return WATCHED_SERVICES[prefix];
  }

  const clean = lower.replace(/[^a-z0-9]/g, '');
  if (WATCHED_SERVICES[clean]) {
    return WATCHED_SERVICES[clean];
  }

  const cleanPrefix = prefix.replace(/[^a-z0-9]/g, '');
  if (WATCHED_SERVICES[cleanPrefix]) {
    return WATCHED_SERVICES[cleanPrefix];
  }

  // Exact or alias/name matches
  for (const meta of Object.values(WATCHED_SERVICES)) {
    const name = (meta.name || '').toLowerCase();
    const aliases = (meta.aliases || []).map((a) => (typeof a === 'string' ? a.toLowerCase() : ''));
    for (const candidate of [name, ...aliases]) {
      if (!candidate) continue;
      if (candidate === lower) return meta;
      const cleanCand = candidate.replace(/[^a-z0-9]/g, '');
      if (clean && clean === cleanCand) return meta;
    }
  }

  // Check if a full canonical name or alias is contained in the query string
  for (const meta of Object.values(WATCHED_SERVICES)) {
    const name = (meta.name || '').toLowerCase();
    const aliases = (meta.aliases || []).map((a) => (typeof a === 'string' ? a.toLowerCase() : ''));
    for (const candidate of [name, ...aliases]) {
      if (candidate && candidate.length >= 4 && lower.includes(candidate)) {
        return meta;
      }
    }
  }

  return null;
}

export function isWatchedApi(apiName: string): boolean {
  return findServiceMeta(apiName) !== null;
}

export function getWatchedApiNames(): string[] {
  return Object.keys(WATCHED_SERVICES).sort();
}

export function getEcosystemForService(serviceOrApi: string): string {
  const meta = findServiceMeta(serviceOrApi);
  return meta?.ecosystem || 'More';
}

export function getCategoryForService(serviceOrApi: string): string | null {
  const meta = findServiceMeta(serviceOrApi);
  return meta?.category || null;
}

export function getReleaseFeedUrls(serviceOrApi: string): string[] {
  const meta = findServiceMeta(serviceOrApi);
  if (!meta) return [];

  const urls: string[] = [];
  if (Array.isArray(meta.release_feed_urls)) {
    for (const u of meta.release_feed_urls) {
      if (u && !urls.includes(u)) {
        urls.push(u);
      }
    }
  }
  if (meta.release_feed_url && !urls.includes(meta.release_feed_url)) {
    urls.push(meta.release_feed_url);
  }
  return urls;
}

export function getReleaseFeedUrl(serviceOrApi: string): string | null {
  const urls = getReleaseFeedUrls(serviceOrApi);
  return urls.length > 0 ? urls[0] : null;
}

export function getOfficialReleaseFeeds(): Record<string, string> {
  const feeds: Record<string, string> = {};
  for (const key of Object.keys(WATCHED_SERVICES)) {
    const urls = getReleaseFeedUrls(key);
    if (urls.length > 0) {
      feeds[key] = urls[0];
    }
  }
  return feeds;
}
